### PR TITLE
Add SymInt::guard_int

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85139

This allows you to explicitly guard on the specific integer value
of a SymInt so that you can condition on it.  If possible, prefer
guarding on a boolean expression instead.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch/csrc/jit/python/init.cpp [hunk 1]
```diff
@@ -156,9 +156,14 @@ class PythonSymIntNodeImpl : public c10::SymIntNodeImpl {
...
 
+  virtual int64_t guard_int(const char* file, int64_t line) override {
+    py::gil_scoped_acquire acquire;
+    return getPyObj().attr("guard_int")(file, line).cast<int64_t>();
+  }
+
   virtual int64_t int_() override {
     py::gil_scoped_acquire acquire;
-    return getPyObj().attr("__int__")().cast<int64_t>();
+    return getPyObj().attr("__int__").cast<int64_t>();
```
Comment (Chillee) [inline torch/csrc/jit/python/init.cpp:166]: This is right?

**Review by Chillee — APPROVED** [review]

File: torch/csrc/jit/python/init.cpp (hunk 1 above)
Comment (ezyang) [reply torch/csrc/jit/python/init.cpp:166]: 🤔 maybe not

**Pull request closed without merging**